src/engines/mod.rs

```diff
@@ -132,22 +132,23 @@ impl<'a, I: 'a + IoProvider> ExecutionState<'a, I> {
             return r;
         }
 
-        // Maybe there's a nicer way to alter the extension without turning
-        // `name` into a Path?
-
-        let mut ext = PathBuf::from(name);
-        let mut ename = OsString::from(match ext.file_name() {
-            Some(s) => s,
-            None => return OpenResult::NotAvailable
-        });
-        ename.push(format_to_extension(format));
-        ext.set_file_name(ename);
-
-        if let FileFormat::Format = format {
-            self.io.input_open_format(&ext.into_os_string(), self.status)
-        } else {
-            self.io.input_open_name(&ext.into_os_string(), self.status)
+        for e in format_to_extension(format) {
+            let mut ext = PathBuf::from(name);
+            if !ext.set_extension(e) {
+                return OpenResult::NotAvailable
+            }
+
+            if let FileFormat::Format = format {
+                if let OpenResult::Ok(r) = self.io.input_open_format(&ext.into_os_string(), self.status) {
```

Comment:
Yup, @pkgw.<sup>*</sup>

###### *Recreating the enum variant inside of the `if let` statement could lead to lost/different type information but in this case its equivalent.